README.md
# Apilytics for Python

[![pypi](https://img.shields.io/pypi/v/apilytics)](https://pypi.org/project/apilytics/)
[![ci](https://github.com/apilytics/apilytics-python/actions/workflows/ci.yml/badge.svg)](https://github.com/apilytics/apilytics-python/actions/workflows/ci.yml)
[![codecov](https://codecov.io/gh/apilytics/apilytics-python/branch/master/graph/badge.svg?token=GIW1NZ7UAJ)](https://codecov.io/gh/apilytics/apilytics-python)
[![mypy checked](http://www.mypy-lang.org/static/mypy_badge.svg)](http://mypy-lang.org)
[![code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![python versions](https://img.shields.io/pypi/pyversions/apilytics)](#what-python-versions-does-the-package-work-with)
[![license](https://img.shields.io/pypi/l/apilytics.svg)](https://github.com/apilytics/apilytics-python/blob/master/LICENSE)

Apilytics is a service that lets you analyze operational, performance and security metrics from your APIs easily.

<img src="https://www.apilytics.io/mock-ups/time-frame.gif" alt="Apilytics dashboard animation" width="600" height="300" />

## Installation

1. Sign up and get your API key from https://apilytics.io - we offer a completely free trial with no credit card required!

2. Install this package:

```sh
pip install apilytics
```

3. Enable the middleware and set your API key:\
   _A good practice is to securely store the API key as an environment variable.
   You can leave the env variable unset in e.g. development and test environments,
   the middleware will be automatically disabled if the key is `None`._

### Django

`settings.py`:

```python
import os

APILYTICS_API_KEY = os.getenv("APILYTICS_API_KEY")

MIDDLEWARE = [
    "apilytics.django.ApilyticsMiddleware",  # Ideally the first middleware in the list.
    # ...
]
```

### FastAPI

`main.py`:

```python
import os

from apilytics.fastapi import ApilyticsMiddleware
from fastapi import FastAPI

app = FastAPI()

# Ideally the last middleware you add.
app.add_middleware(ApilyticsMiddleware, api_key=os.getenv("APILYTICS_API_KEY"))
```

### Flask

`app.py`:

```python
import os

from apilytics.flask import apilytics_middleware
from flask import Flask

app = Flask(__name__)

# Ideally wrap your app with the middleware before you do anything else with it.
app = apilytics_middleware(app, api_key=os.getenv("APILYTICS_API_KEY"))
```

### Other Python Frameworks

You can easily build your own middleware which measures the execution time and sends the metrics:

`my_apilytics_middleware.py`:

```python
import os

from apilytics.core import ApilyticsSender


def my_apilytics_middleware(request, get_response):
    api_key = os.getenv("APILYTICS_API_KEY")
    if not api_key:
        return get_response(request)

    with ApilyticsSender(
        api_key=api_key,
        path=request.path,
        query=request.query_string,
        method=request.method,
        request_size=len(request.body),
        user_agent=request.headers.get("user-agent"),
        ip=request.headers.get("x-forwarded-for", "").split(",")[0].strip(),
    ) as sender:
        response = get_response(request)
        sender.set_response_info(
            status_code=response.status_code,
            response_size=len(response.body),
        )
    return response
```

## Frequently Asked Questions

### Does the middleware slow down my backend?

- No. The middleware does all of its requests to the Apilytics API in a background thread pool,
  so it will not slow down your normal request handling.

### What 3rd party dependencies does `apilytics` have?

- None besides the frameworks that you use it in.

### What Python versions does the package work with?

- `apilytics` is tested to work on all the currently [supported versions of Python](https://devguide.python.org/#status-of-python-branches): 3.7, 3.8, 3.9, and 3.10.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.6.0] - 2022-03-26

### Added

- Add `apilytics.flask.apilytics_middleware` for sending metrics from Flask applications.

### Fixed

- Fix FastAPI middleware suggested installation order; it should ideally be the *last* one added.

## [1.5.0] - 2022-03-17

### Added

- Send user's IP address with metrics. Used for visualization aggregate geolocation data.
  The IP is never stored, and it is never sent to 3rd parties.

## [1.4.0] - 2022-02-20

### Added

- Send current system CPU usage together with metrics on Linux systems.
- Send current system memory usage and total available memory together with metrics on Linux systems.
- Add platform name to sent Apilytics version info.

## [1.3.0] - 2022-02-02

### Added

- Send `User-Agent` information with metrics.

## [1.2.1] - 2022-02-01

### Fixed

- Don't send 0 as the response size for streaming responses, send nothing instead.

## [1.2.0] - 2022-01-31

### Added

- Send request and response body size information with metrics.

### Changed

- Change `status_code` into an optional parameter in `ApilyticsSender.set_response_info`.

## [1.1.0] - 2022-01-16

### Added

- Send query parameters in addition to the path.

## [1.0.2] - 2022-01-16

### Added

- Send Apilytics version info together with metrics.

## [1.0.1] - 2022-01-12

### Fixed

- Improve README documentation.

## [1.0.0] - 2022-01-10

### Added

- Initial version with Django and FastAPI support.


tests/conftest.py
"""General helpers that can be used in any tests."""
import concurrent.futures
import json
import unittest.mock
from typing import Any, Callable, Generator, TypeVar

import pytest

__all__ = ["decode_request_data"]

T = TypeVar("T")


def decode_request_data(data: bytes) -> Any:
    return json.loads(data.decode())


@pytest.fixture
def mocked_urlopen() -> Generator[unittest.mock.MagicMock, None, None]:
    with unittest.mock.patch("apilytics.core.urllib.request.urlopen") as mocked:
        yield mocked


class _MockedExecutor(concurrent.futures.ThreadPoolExecutor):
    # Ignore: Parent is typed to return a `Future[T]`.
    def submit(self, fn: Callable[..., T], *args: Any, **kwargs: Any) -> T:  # type: ignore[override]
        """Overridden to make this synchronous during tests."""
        return super().submit(fn, *args, **kwargs).result()


@pytest.fixture(scope="session", autouse=True)
def mocked_executor() -> Generator[None, None, None]:
    with unittest.mock.patch(
        "apilytics.core.concurrent.futures.ThreadPoolExecutor",
        new=_MockedExecutor,
    ):
        yield


@pytest.fixture(scope="session", autouse=True)
def mocked_sleep() -> Generator[None, None, None]:
    with unittest.mock.patch("apilytics.core.time.sleep", new=lambda secs: None):
        yield


tests/flask/app.py
"""Simple Flask app that's used in tests."""
from typing import NoReturn

import flask
import flask.typing

import apilytics.flask

app = flask.Flask(__name__)

app = apilytics.flask.apilytics_middleware(app, api_key="dummy-key")


@app.get("/error")
def error_route() -> NoReturn:
    raise RuntimeError


@app.post("/empty")
def no_body_route() -> flask.typing.ResponseReturnValue:
    return "", 200


@app.get("/streaming")
def streaming_route() -> flask.typing.ResponseReturnValue:
    return app.response_class(
        iter([b"first", b" ", b"second"]), mimetype="application/octet-stream"
    )


@app.route("/", methods=["GET", "POST"], defaults={"path": ""})
@app.route("/<path:path>", methods=["GET", "POST"])
def ok_route(path: str) -> flask.typing.ResponseReturnValue:
    if flask.request.method == "POST":
        return b"created", 201
    return b"ok", 200


tests/test_core.py
import platform
import sys
import textwrap
import time
import unittest.mock
import urllib.error

import apilytics.core
import tests.conftest


# Restore the real sleep behavior for this one test for thoroughness.
@unittest.mock.patch("apilytics.core.time.sleep", new=time.sleep)
def test_apilytics_sender_should_call_apilytics_api(
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    with apilytics.core.ApilyticsSender(
        api_key="dummy-key",
        path="/",
        method="GET",
    ) as sender:
        sender.set_response_info(status_code=200)

    assert mocked_urlopen.call_count == 1

    call_args, call_kwargs = mocked_urlopen.call_args
    assert not call_args
    assert call_kwargs.keys() == {"url", "data"}

    api_request = call_kwargs["url"]

    assert api_request.full_url == "https://www.apilytics.io/api/v1/middleware"
    assert api_request.method == "POST"
    assert api_request.headers == {
        # urllib calls `capitalize()` on the header keys.
        "Content-type": "application/json",
        "X-api-key": "dummy-key",
        "Apilytics-version": f"apilytics-python-core/{apilytics.__version__};python/{platform.python_version()};;{sys.platform}",
    }

    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert data.keys() == {
        "path",
        "method",
        "statusCode",
        "timeMillis",
        *(
            ("cpuUsage", "memoryUsage", "memoryTotal")
            if platform.system() == "Linux"
            else ()
        ),
    }
    assert data["path"] == "/"
    assert data["method"] == "GET"
    assert data["statusCode"] == 200
    assert isinstance(data["timeMillis"], int)
    if platform.system() == "Linux":
        assert 0 <= data["cpuUsage"] <= 1
        assert data["memoryUsage"] > 0
        assert data["memoryTotal"] > data["memoryUsage"]


def test_apilytics_sender_should_send_query_params(
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    with apilytics.core.ApilyticsSender(
        api_key="dummy-key",
        path="/path",
        query="key=value?other=123",
        method="PUT",
    ) as sender:
        sender.set_response_info(status_code=200)

    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert data["path"] == "/path"
    assert data["query"] == "key=value?other=123"


def test_apilytics_sender_should_not_send_empty_query_params(
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    with apilytics.core.ApilyticsSender(
        api_key="dummy-key",
        path="/",
        query="",
        method="GET",
    ) as sender:
        sender.set_response_info(status_code=200)

    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert "query" not in data

    with apilytics.core.ApilyticsSender(
        api_key="dummy-key",
        path="/",
        query=None,
        method="GET",
    ) as sender:
        sender.set_response_info(status_code=200)

    assert mocked_urlopen.call_count == 2
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert "query" not in data


def test_apilytics_sender_should_handle_empty_values_correctly(
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    with apilytics.core.ApilyticsSender(
        api_key="dummy-key",
        path="",
        method="",
        query="",
        request_size=None,
        user_agent=None,
        apilytics_integration=None,
        integrated_library=None,
    ) as sender:
        sender.set_response_info(status_code=None, response_size=None)

    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert data.keys() == {
        "path",
        "method",
        "timeMillis",
        *(
            ("cpuUsage", "memoryUsage", "memoryTotal")
            if platform.system() == "Linux"
            else ()
        ),
    }
    assert data["path"] == ""
    assert data["method"] == ""
    assert isinstance(data["timeMillis"], int)
    if platform.system() == "Linux":
        assert isinstance(data["cpuUsage"], float)
        assert isinstance(data["memoryUsage"], int)
        assert isinstance(data["memoryTotal"], int)


@unittest.mock.patch("apilytics.core.platform.system", return_value="Linux")
@unittest.mock.patch("apilytics.core._get_used_and_total_memory", return_value=(0, 0))
def test_apilytics_sender_should_read_proc_stat_on_linux(
    _mocked_system: unittest.mock.MagicMock,
    _mocked_memory: unittest.mock.MagicMock,
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    mocked_stat_start = textwrap.dedent(
        """\
        cpu  27133 0 33621 13668027 1459 0 508 10 100 100
        cpu0 7260 0 7391 3420134 484 0 80 0 0 0
        cpu1 7346 0 9306 3412338 138 0 82 0 0 0
        """  # The real file is longer.
    )
    mocked_stat_end = textwrap.dedent(
        """\
        cpu  28869 0 33657 13680890 1460 0 508 20 200 200
        cpu0 7263 0 7398 3423775 484 0 80 0 0 0
        cpu1 9069 0 9314 3414266 138 0 82 0 0 0
        """
    )
    with unittest.mock.patch(
        "builtins.open",
        new=unittest.mock.mock_open(read_data=mocked_stat_start),
    ) as mocked_open:
        mocked_open.side_effect = (
            mocked_open.return_value,
            unittest.mock.mock_open(read_data=mocked_stat_end).return_value,
        )
        with apilytics.core.ApilyticsSender(
            api_key="dummy-key",
            path="/",
            method="GET",
        ) as sender:
            sender.set_response_info(status_code=200)

    assert mocked_open.call_count == 2
    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    # Totals ignore last two "guest" fields.
    total_start = 27133 + 0 + 33621 + 13668027 + 1459 + 0 + 508 + 10
    total_end = 28869 + 0 + 33657 + 13680890 + 1460 + 0 + 508 + 20
    # Idles combine idle and iowait.
    idle_start = 13668027 + 1459
    idle_end = 13680890 + 1460
    assert data["cpuUsage"] == 1 - (idle_end - idle_start) / (total_end - total_start)
    assert data["cpuUsage"] == 0.12167144612863579


@unittest.mock.patch("apilytics.core.platform.system", return_value="Linux")
@unittest.mock.patch("apilytics.core._get_used_and_total_memory", return_value=(0, 0))
def test_apilytics_sender_should_handle_proc_stat_read_failure(
    _mocked_system: unittest.mock.MagicMock,
    _mocked_memory: unittest.mock.MagicMock,
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    with unittest.mock.patch("builtins.open", side_effect=OSError) as mocked_open:
        with apilytics.core.ApilyticsSender(
            api_key="dummy-key",
            path="/",
            method="GET",
        ) as sender:
            sender.set_response_info(status_code=200)

    assert mocked_open.call_count == 1
    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert "cpuUsage" not in data


@unittest.mock.patch("apilytics.core.platform.system", return_value="Linux")
@unittest.mock.patch("apilytics.core._get_used_and_total_memory", return_value=(0, 0))
def test_apilytics_sender_should_handle_proc_stat_iowait_missing(
    _mocked_system: unittest.mock.MagicMock,
    _mocked_memory: unittest.mock.MagicMock,
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    mocked_stat_start = "cpu  27133 0 33621 13668027"
    mocked_stat_end = "cpu  28869 0 33657 13680890"
    with unittest.mock.patch(
        "builtins.open",
        new=unittest.mock.mock_open(read_data=mocked_stat_start),
    ) as mocked_open:
        mocked_open.side_effect = (
            mocked_open.return_value,
            unittest.mock.mock_open(read_data=mocked_stat_end).return_value,
        )
        with apilytics.core.ApilyticsSender(
            api_key="dummy-key",
            path="/",
            method="GET",
        ) as sender:
            sender.set_response_info(status_code=200)

    assert mocked_open.call_count == 2
    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    total_start = 27133 + 0 + 33621 + 13668027
    total_end = 28869 + 0 + 33657 + 13680890
    idle_start = 13668027
    idle_end = 13680890
    assert data["cpuUsage"] == 1 - (idle_end - idle_start) / (total_end - total_start)
    assert data["cpuUsage"] == 0.12107960368978476


@unittest.mock.patch("apilytics.core.platform.system", return_value="Linux")
@unittest.mock.patch("apilytics.core._get_used_and_total_memory", return_value=(0, 0))
def test_apilytics_sender_should_handle_proc_stat_timers_not_increased_zero_division(
    _mocked_system: unittest.mock.MagicMock,
    _mocked_memory: unittest.mock.MagicMock,
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    mocked_stat = "cpu  27133 0 33621 13668027"
    with unittest.mock.patch(
        "builtins.open",
        new=unittest.mock.mock_open(read_data=mocked_stat),
    ) as mocked_open:
        mocked_open.side_effect = (
            mocked_open.return_value,
            unittest.mock.mock_open(read_data=mocked_stat).return_value,
        )
        with apilytics.core.ApilyticsSender(
            api_key="dummy-key",
            path="/",
            method="GET",
        ) as sender:
            sender.set_response_info(status_code=200)

    assert mocked_open.call_count == 2
    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert data["cpuUsage"] == 0.0


@unittest.mock.patch("apilytics.core.platform.system", return_value="Windows")
@unittest.mock.patch("apilytics.core._get_used_and_total_memory", return_value=(0, 0))
def test_apilytics_sender_should_not_read_proc_stat_when_not_on_linux(
    _mocked_system: unittest.mock.MagicMock,
    _mocked_memory: unittest.mock.MagicMock,
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    with unittest.mock.patch("builtins.open") as mocked_open:
        with apilytics.core.ApilyticsSender(
            api_key="dummy-key",
            path="/",
            method="GET",
        ) as sender:
            sender.set_response_info(status_code=200)

    assert mocked_open.call_count == 0
    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert "cpuUsage" not in data


@unittest.mock.patch("apilytics.core.platform.system", return_value="Linux")
@unittest.mock.patch("apilytics.core._get_cpu_usage", return_value=0.0)
def test_apilytics_sender_should_read_proc_meminfo_on_linux(
    _mocked_system: unittest.mock.MagicMock,
    _mocked_cpu_usage: unittest.mock.MagicMock,
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    memory_total = 4_125_478_912
    memory_available = 3_360_526_336

    mocked_meminfo = textwrap.dedent(
        f"""\
        MemTotal:        {memory_total // 1024} kB
        MemFree:          789940 kB
        MemAvailable:    {memory_available // 1024} kB
        Buffers:         2450168 kB
        """  # The real file is longer.
    )
    with unittest.mock.patch(
        "builtins.open", new=unittest.mock.mock_open(read_data=mocked_meminfo)
    ) as mocked_open:
        with apilytics.core.ApilyticsSender(
            api_key="dummy-key",
            path="/",
            method="GET",
        ) as sender:
            sender.set_response_info(status_code=200)

    assert mocked_open.call_count == 1
    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert data["memoryUsage"] == memory_total - memory_available
    assert data["memoryTotal"] == memory_total


@unittest.mock.patch("apilytics.core.platform.system", return_value="Linux")
@unittest.mock.patch("apilytics.core._get_cpu_usage", return_value=0.0)
def test_apilytics_sender_should_handle_proc_meminfo_read_failure(
    _mocked_system: unittest.mock.MagicMock,
    _mocked_cpu_usage: unittest.mock.MagicMock,
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    with unittest.mock.patch("builtins.open", side_effect=OSError) as mocked_open:
        with apilytics.core.ApilyticsSender(
            api_key="dummy-key",
            path="/",
            method="GET",
        ) as sender:
            sender.set_response_info(status_code=200)

    assert mocked_open.call_count == 1
    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert "memoryUsage" not in data
    assert "memoryTotal" not in data


@unittest.mock.patch("apilytics.core.platform.system", return_value="Linux")
@unittest.mock.patch("apilytics.core._get_cpu_usage", return_value=0.0)
def test_apilytics_sender_should_handle_proc_meminfo_total_missing(
    _mocked_system: unittest.mock.MagicMock,
    _mocked_cpu_usage: unittest.mock.MagicMock,
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    with unittest.mock.patch(
        "builtins.open", new=unittest.mock.mock_open(read_data="")
    ) as mocked_open:
        with apilytics.core.ApilyticsSender(
            api_key="dummy-key",
            path="/",
            method="GET",
        ) as sender:
            sender.set_response_info(status_code=200)

    assert mocked_open.call_count == 1
    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert "memoryUsage" not in data
    assert "memoryTotal" not in data


@unittest.mock.patch("apilytics.core.platform.system", return_value="Linux")
@unittest.mock.patch("apilytics.core._get_cpu_usage", return_value=0.0)
def test_apilytics_sender_should_handle_proc_meminfo_available_missing(
    _mocked_system: unittest.mock.MagicMock,
    _mocked_cpu_usage: unittest.mock.MagicMock,
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    memory_total = 1048576
    with unittest.mock.patch(
        "builtins.open",
        new=unittest.mock.mock_open(read_data=f"MemTotal: {memory_total // 1024}"),
    ) as mocked_open:
        with apilytics.core.ApilyticsSender(
            api_key="dummy-key",
            path="/",
            method="GET",
        ) as sender:
            sender.set_response_info(status_code=200)

    assert mocked_open.call_count == 1
    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert "memoryUsage" not in data
    assert data["memoryTotal"] == memory_total


@unittest.mock.patch("apilytics.core.platform.system", return_value="Windows")
@unittest.mock.patch("apilytics.core._get_cpu_usage", return_value=0.0)
def test_apilytics_sender_should_not_read_proc_meminfo_when_not_on_linux(
    _mocked_system: unittest.mock.MagicMock,
    _mocked_cpu_usage: unittest.mock.MagicMock,
    mocked_urlopen: unittest.mock.MagicMock,
) -> None:
    with unittest.mock.patch("builtins.open") as mocked_open:
        with apilytics.core.ApilyticsSender(
            api_key="dummy-key",
            path="/",
            method="GET",
        ) as sender:
            sender.set_response_info(status_code=200)

    assert mocked_open.call_count == 0
    assert mocked_urlopen.call_count == 1
    __, call_kwargs = mocked_urlopen.call_args
    data = tests.conftest.decode_request_data(call_kwargs["data"])
    assert "memoryUsage" not in data
    assert "memoryTotal" not in data


@unittest.mock.patch(
    "apilytics.core.urllib.request.urlopen",
    side_effect=urllib.error.URLError("testing"),
)
def test_apilytics_sender_should_hide_http_errors(
    mocked_erroring_urlopen: unittest.mock.MagicMock,
) -> None:
    with apilytics.core.ApilyticsSender(
        api_key="dummy-key",
        path="/",
        method="GET",
    ) as sender:
        sender.set_response_info(status_code=200)

    assert mocked_erroring_urlopen.call_count == 1


apilytics/core.py
import concurrent.futures
import json
import platform
import re
import sys
import time
import types
import urllib.error
import urllib.request
from typing import ClassVar, Optional, Tuple, Type

import apilytics

__all__ = ["ApilyticsSender"]


class ApilyticsSender:
    """
    Context manager that sends API analytics data to Apilytics (https://apilytics.io).

    Use as a context manager by wrapping a request handler in an HTTP middleware.
    Make sure to call ``set_response_info`` on the object received in the with-statement.

    The metrics will be sent in a fire-and-forget background task when the block ends.

    Examples:
        my_middleware.py::

            with ApilyticsSender(
                api_key="<your-api-key>",
                path=request.path,
                query=request.query_string,
                method=request.method,
                request_size=len(request.body),
                user_agent=request.headers.get("user-agent"),
                ip=request.headers.get("x-forwarded-for", "").split(",")[0].strip(),
            ) as sender:
                response = get_response(request)
                sender.set_response_info(
                    status_code=response.status_code,
                    response_size=len(response.body),
                )
    """

    _executor: ClassVar[concurrent.futures.Executor]

    _apilytics_version_template: ClassVar[
        str
    ] = f"{{integration}}/{apilytics.__version__};python/{platform.python_version()};{{library}};{sys.platform}"

    def __init__(
        self,
        *,
        api_key: str,
        path: str,
        method: str,
        query: Optional[str] = None,
        request_size: Optional[int] = None,
        user_agent: Optional[str] = None,
        ip: Optional[str] = None,
        apilytics_integration: Optional[str] = None,
        integrated_library: Optional[str] = None,
        prevent_send_on_exit: bool = False,
    ) -> None:
        """
        Initialize the context manager with info from the HTTP request object.

        Args:
            api_key: The API key for your Apilytics origin.
            path: Path of the user's HTTP request, e.g. "/foo/bar/123".
            method: Method of the user's HTTP request, e.g. "GET".
            query: Optional query string of the user's HTTP request e.g. "key=val&other=123".
                An empty string and None are treated equally. Can have an optional "?" at the start.
            request_size: Size of the user's HTTP request's body in bytes.
            user_agent: Value of the `User-Agent` header from the user's HTTP request.
                An empty string and None are treated equally.
            ip: User's IP address (used for geolocation, never stored nor sent to 3rd parties).
                An empty string and None are treated equally.
            apilytics_integration: Name of the Apilytics integration that's calling this,
                e.g. "apilytics-python-django". No need to pass this when calling from user code.
            integrated_library: Name and version of the integration that this is used in,
                e.g. "django/3.2.1". No need to pass this when calling from user code.
            prevent_send_on_exit: Don't immediately send the metrics when the context
                manager exits. Useful for advanced deferred sending scenarios.
        """
        self._api_key = api_key
        self._path = path
        self._method = method
        self._query = query
        self._request_size = request_size
        self._user_agent = user_agent
        self._ip = ip

        self._response_size: Optional[int] = None
        self._status_code: Optional[int] = None

        self._apilytics_version = self._apilytics_version_template.format(
            integration=apilytics_integration or "apilytics-python-core",
            library=integrated_library or "",
        )

        self._prevent_send_on_exit = prevent_send_on_exit

    def __enter__(self) -> "ApilyticsSender":
        """Start the timer, measuring how long the ``with`` block takes to execute."""
        self._start_time_ns = time.perf_counter_ns()
        return self

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_val: Optional[BaseException],
        exc_tb: Optional[types.TracebackType],
    ) -> None:
        if self._prevent_send_on_exit:
            return
        self.send()

    def set_response_info(
        self, *, status_code: Optional[int] = None, response_size: Optional[int] = None
    ) -> None:
        """
        Update the context manager with info from the HTTP response object.

        Should be called before the context manager's block ends.

        Args:
            status_code: Status code for the HTTP response. Can be omitted (or None)
                if the middleware could not get the status code.
            response_size: Size of the sent HTTP response's body in bytes.
        """
        self._status_code = status_code
        self._response_size = response_size

    def send(self) -> None:
        """Send metrics to Apilytics in a fire-and-forget background task."""
        self._end_time_ns = time.perf_counter_ns()
        if not hasattr(self, "_executor"):
            # Use only a single background thread and share the pool to minimize
            # resource hogging.
            self.__class__._executor = concurrent.futures.ThreadPoolExecutor(
                max_workers=1
            )
        self._executor.submit(self._send_metrics)

    def _send_metrics(self) -> None:
        memory_usage, memory_total = _get_used_and_total_memory()
        cpu_usage = _get_cpu_usage()

        request = urllib.request.Request(
            url="https://www.apilytics.io/api/v1/middleware",
            method="POST",
            headers={
                "Content-Type": "application/json",
                "X-API-Key": self._api_key,
                "Apilytics-Version": self._apilytics_version,
            },
        )
        data = {
            "path": self._path,
            "method": self._method,
            "timeMillis": (self._end_time_ns - self._start_time_ns) // 1_000_000,
            # Don't send empty strings.
            **({"query": self._query} if self._query else {}),
            **({"userAgent": self._user_agent} if self._user_agent else {}),
            **({"ip": self._ip} if self._ip else {}),
            **(
                {"statusCode": self._status_code}
                if self._status_code is not None
                else {}
            ),
            **(
                {"requestSize": self._request_size}
                if self._request_size is not None
                else {}
            ),
            **(
                {"responseSize": self._response_size}
                if self._response_size is not None
                else {}
            ),
            **({"cpuUsage": cpu_usage} if cpu_usage is not None else {}),
            **({"memoryUsage": memory_usage} if memory_usage is not None else {}),
            **({"memoryTotal": memory_total} if memory_total is not None else {}),
        }
        try:
            urllib.request.urlopen(url=request, data=json.dumps(data).encode())
        except urllib.error.URLError:
            pass


def _get_cpu_usage() -> Optional[float]:
    """
    Get the current CPU usage as a percentage.

    Returns:
        A percentage value between 0 and 1, or None if the CPU usage could not be
        determined (most likely because the system is not Linux).
    """
    if platform.system() != "Linux":
        return None

    def cpu_times() -> Tuple[int, int]:
        with open("/proc/stat") as f:
            stat = f.readline()

        # Ignore the `cpu` text from the start and the last two "guest" times.
        times = [int(val) for val in stat.split()[1:9]]

        total = sum(times)
        idle = times[3]

        try:
            # Include `iowait` time into idle time if available, as does:
            # https://github.com/torvalds/linux/blob/4f12b742eb2b3a850ac8be7dc4ed52976fc6cb0b/kernel/sched/cputime.c#L225
            idle += times[4]
        except IndexError:
            # `iowait` time is not available before Linux 2.5.41, quite unlikely
            # to happen but doesn't hurt to handle this anyway.
            pass

        return idle, total

    try:
        idle_start, total_start = cpu_times()

        # There is no such thing as CPU usage percentage on a single point of time.
        # At any discrete instant a CPU core is either fully used or fully idle.
        # This is why we need to measure the usage over a known time interval. An
        # interval of one second has been tested to provide quite consistent results.
        time.sleep(1)

        idle_end, total_end = cpu_times()
    except OSError:
        return None

    try:
        idle_percentage = (idle_end - idle_start) / (total_end - total_start)
    except ZeroDivisionError:
        return 0.0

    return 1 - idle_percentage


def _get_used_and_total_memory() -> Tuple[Optional[int], Optional[int]]:
    """
    Get information about the used and total system memory.

    Returns:
        A tuple containing the used and total system memory in bytes.
        (None, None) if the system is not Linux or if the reading fails.
        (None, int) if the used memory could not be determined.
    """
    used = None
    total = None

    if platform.system() == "Linux":
        try:
            with open("/proc/meminfo") as f:
                meminfo = f.read()
        except OSError:
            pass  # Prepare for everything and anything.
        else:
            total_match = re.search(r"MemTotal:\s*(\d+)", meminfo)
            available_match = re.search(r"MemAvailable:\s*(\d+)", meminfo)
            if total_match:
                total = int(total_match.group(1)) * 1024
                if available_match:
                    # If MemAvailable exists MemTotal will also exist.
                    # The reverse is not always true (MemAvailable came in Linux 3.14).
                    available = int(available_match.group(1)) * 1024
                    used = total - available

    return used, total
